src/Interpreters/ClusterProxy/executeQuery.cpp

```diff
@@ -861,6 +898,207 @@ bool canUseParallelReplicasOnInitiator(const ContextPtr & context)
+        }
+
+        return false;
+    };
+
+    return is_reading_with_parallel_replicas(plan.getRootNode());
+}
+
+/// find and remove ReadFromParallelRemoteReplicasStep in query plan,
+/// also returns parallel replicas coordinator stored in ReadFromParallelRemoteReplicasStep
+ParallelReplicasReadingCoordinatorPtr dropReadFromRemoteInPlan(QueryPlan & query_plan)
+{
+    struct Frame
+    {
+        QueryPlan::Node * node = nullptr;
+        QueryPlan::Node * parent = nullptr;
+        size_t next_child = 0;
+    };
+
+    using Stack = std::vector<Frame>;
+
+    Stack stack;
+    stack.push_back(Frame{.node = query_plan.getRootNode()});
+
+    while (!stack.empty())
+    {
+        auto & frame = stack.back();
+
+        if (frame.node->children.empty())
+        {
+            const auto * step = frame.node->step.get();
+            if (const auto * read_from_remote = typeid_cast<const ReadFromParallelRemoteReplicasStep *>(step))
+            {
+                auto & children = frame.parent->children;
+                for (auto it = begin(children); it != end(children); ++it)
+                {
+                    if ((*it)->step.get() == step)
+                    {
+                        children.erase(it);
+                        return read_from_remote->getCoordinator();
+                    }
+                }
+            }
+        }
+
+        /// Traverse all children first.
+        if (frame.next_child < frame.node->children.size())
+        {
+            auto next_frame = Frame{.node = frame.node->children[frame.next_child], .parent = frame.node};
+            ++frame.next_child;
+            stack.push_back(next_frame);
+            continue;
+        }
+
+        stack.pop_back();
+    }
+
+    return nullptr;
+}
+
+std::optional<QueryPipeline> executeInsertSelectWithParallelReplicas(
+    const ASTInsertQuery & query_ast,
+    const ContextPtr & context,
+    std::optional<QueryPipeline> local_pipeline,
+    std::optional<ParallelReplicasReadingCoordinatorPtr> coordinator)
+{
+    auto logger = getLogger("executeInsertSelectWithParallelReplicas");
+    LOG_DEBUG(logger, "Executing query with parallel replicas: {}", query_ast.formatForLogging());
+
+    const auto & settings = context->getSettingsRef();
+
+    auto new_context = updateContext(logger, context);
+    auto [cluster, shard_num] = prepairCluster(logger, new_context);
+    auto [connection_pools, max_replicas_to_use] = prepairConnectionPools(logger, new_context, cluster);
+    std::optional<size_t> local_replica_index;
+
+    if (coordinator)
+    {
+        chassert(local_pipeline);
+
+        local_replica_index = findLocalReplicaIndexAndUpdatePools(connection_pools, max_replicas_to_use, cluster);
+
+        /// with local pipeline the snapshot replica is local one
+        /// and replica number was already assigned to it
+        /// so, we need to reuse it here
+        auto snapshot_replica_num = (*coordinator)->getSnapshotReplicaNum();
```

Comment:
Maybe we should add a comment on how it is possible, i.e. what is so special about INSERT SELECT with parallel replicas